test/common/router/router_fuzz_test.cc

```diff
@@ -0,0 +1,482 @@
+                         -> Http::ConnectionPool::Cancellable* {
+                response_decoder = &decoder;
+                // still not too sure what these callbacks are for / do
+                callbacks.onPoolReady(encoder1, cm_.conn_pool_.host_, upstream_stream_info_);
+                return nullptr;
+              }));
+
+      // this creates the response timeout timer
+      expectResponseTimerCreate();
+
+      Http::TestRequestHeaderMapImpl headers{{"x-envoy-retry-on", "5xx"},
+                                             {"x-envoy-internal", "true"},
+                                             {"x-envoy-upstream-rq-timeout-ms", "200"}};
+      /* {"x-envoy-upstream-rq-per-try-timeout-ms", "100"}}; */
+      HttpTestUtility::addDefaultHeaders(headers);
+      router_.decodeHeaders(headers, true);
+      // that was all the set up code to get the router to send a request to the server, now we can
+      // perform interesting actions like getting the request to timeout so the router has to retry,
+      // etc.
+
+      switch (action) {
+      case test::common::router::RouterTestCase::RETRY: {
+
+        // send a 5xx response so the router retries, since the retry-on header is present
+        router_.retry_state_->expectHeadersRetry();
+        Http::ResponseHeaderMapPtr response_headers1(
+            new Http::TestResponseHeaderMapImpl{{":status", "503"}});
+        // simulate receiving a 503 status from the server with the response decoder
+        response_decoder->decodeHeaders(std::move(response_headers1), true);
+
+        // the response should cause the router to make a new request
+        // not too sure why there's a new encoder
+        NiceMock<Http::MockRequestEncoder> encoder2;
+        EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+            .WillOnce(Invoke(
+                [&](Http::ResponseDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                    -> Http::ConnectionPool::Cancellable* {
+                  response_decoder = &decoder;
+                  callbacks.onPoolReady(encoder2, cm_.conn_pool_.host_, upstream_stream_info_);
+                  return nullptr;
+                }));
+        router_.retry_state_->callback_();
+
+        // seems like this is hardcoding in that the router should not retry, since
+        // expectHeadersRetry() above sets the expectation that this returns RetryStatus::Yes, so is
+        // this actually testing that the router automatically retries?
+        EXPECT_CALL(*router_.retry_state_, shouldRetryHeaders(_, _))
+            .WillOnce(Return(RetryStatus::No));
+        Http::ResponseHeaderMapPtr response_headers2(
+            new Http::TestResponseHeaderMapImpl{{":status", "200"}});
+        // simulate a normal response from the server so the router should not retry now
+        response_decoder->decodeHeaders(std::move(response_headers2), true);
```

Comment:
hm probably want two structures -- a request direction (for router_.decodeHeaders) and a response direction (for response_decoder)

Reply:
Sorry what do you mean by two structures?

Reply:
i.e. the input protobuf might look like 
message Action {
  headers
  data
  trailers
}

message Actions {
    Action respnse_action;
    Action request_action;
    google.protobuf.Empty advance_time;
    // etc
}

Reply:
Ah right ok I see thanks